guests/python/src/lib.rs

```diff
@@ -3,82 +3,227 @@
+            .into_iter()
+            .map(|column_value| column_value.to_array(number_rows))
+            .collect::<Result<Vec<_>, _>>()?;
 
-            let locals = [("os", py.import("os")?)].into_py_dict(py)?;
-            let code = c_str!("os.getenv('USER') or os.getenv('USERNAME') or 'Unknown'");
-            let user: String = py.eval(code, None, Some(&locals))?.extract()?;
+        Python::with_gil(|py| {
+            let mut parameter_iters = arrays
+                .iter()
+                .zip(&self.python_function.signature.parameters)
+                .map(|(array, t)| t.arrow_to_python(array, py))
+                .collect::<Result<Vec<_>, _>>()?;
 
-            let s = format!("Hello {user}, I'm Python {version}");
-            PyResult::Ok(s)
-        })
-        .map_err(|e| DataFusionError::External(Box::new(e)))?;
+            let handle = self.python_function.handle.bind(py);
+            let mut output_row_builder = self
+                .python_function
+                .signature
+                .return_type
+                .python_to_arrow(py, number_rows);
+
+            for _ in 0..number_rows {
+                // poll ALL iterators before evaluating the controlflow
+                let params = parameter_iters
+                    .iter_mut()
+                    .map(|it| it.next().expect("all iterators have n_rows"))
+                    .collect::<Result<Vec<_>, _>>()?;
+
+                // determine if we shall call the Python method or not
+                let maybe_params = params.into_iter().try_fold(
+                    Vec::with_capacity(parameter_iters.len()),
+                    |mut accu, next| {
+                        accu.push(next?);
+                        ControlFlow::Continue(accu)
+                    },
+                );
 
-        Ok(ColumnarValue::Scalar(ScalarValue::Utf8(Some(s))))
+                match maybe_params {
+                    ControlFlow::Continue(params) => {
+                        let params = PyTuple::new(py, params).map_err(|e| {
+                            exec_datafusion_err!(
+                                "cannot create parameter tuple: {}",
+                                py_err_to_string(e, py)
+                            )
+                        })?;
+                        let rval = handle.call1(params).map_err(|e| {
+                            exec_datafusion_err!(
+                                "cannot call function: {}",
+                                py_err_to_string(e, py)
+                            )
+                        })?;
+                        output_row_builder.push(rval)?;
+                    }
+                    ControlFlow::Break(()) => {
+                        output_row_builder.skip();
+                    }
+                }
+            }
+
+            // check invariants
+            for mut it in parameter_iters {
+                assert!(it.next().is_none(), "iterator should be done");
```

Comment:
It might be useful to do like
```suggestion
                let next = it.next();
                assert!(next.is_none(), "iterator should be done but produced {next}");
```

(though I doubt `next` impls `Display` so this exact code probably won't work) just so we have more insight into what exactly might be failing if we run into this.

Also, I feel like it's somewhat unclear where this invariant is upheld - is it up above when we check the `args.len()` against the signature parameters len or something like that?